### PR TITLE
DLPX-66733 zfs share -g should ignore ZFS_CANMOUNT_NOAUTO

### PR DESCRIPTION
**Pull request by don-brady**
### Motivation and Context
The `zfs share -g` command will skip nfs exports that have their `canmount` property set to `noauto` but otherwise have a valid mountpoint.  For NPM filesystems the `canmount`  and `sync` properties are overridden. This causes the `zfs-share` service to drop their exports.

### How Has This Been Tested?
git ab-pre-push: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/2529/

Also installed the fix onto an engine that had the condition and the `zfs share -g` now correctly ignores `canmount=noauto` but it is mounted when filtering the datasets to export.

### CODE REVIEW

File: cmd/zfs/zfs_main.c [hunk 1]
```diff
@@ -6549,7 +6549,7 @@ share_mount_one(zfs_handle_t *zhp, int op, int flags, char *protocol,
 		    "'canmount' property is set to 'off'\n"), cmdname,
 		    zfs_get_name(zhp));
 		return (1);
-	} else if (canmount == ZFS_CANMOUNT_NOAUTO && !explicit) {
+	} else if (canmount == ZFS_CANMOUNT_NOAUTO && !explicit && !generate) {
```
Comment (pzakha) [inline cmd/zfs/zfs_main.c]: Should we also check that the file system is already mounted?

Also I've read the bug and there was an open question as to why those datasets were set to `noauto`. What was the reasoning there?

Comment (grwilson) [reply cmd/zfs/zfs_main.c]: Agreed, I think we only want to allow the share to get generated if it's mounted.

Comment (don-brady) [reply cmd/zfs/zfs_main.c]: The `mountpoint` property is checked just before checking `canmount` in `share_mount_one()`.

Comment (pzakha) [reply cmd/zfs/zfs_main.c]: There's a `mounted` property too, which can be set to 'no' if the dataset is not mounted. It is different from `mountpoint`. See for instance:
```$ zfs get -p mountpoint,canmount,mounted rpool/fs1
NAME       PROPERTY    VALUE       SOURCE
rpool/fs1  mountpoint  /mnt/fs1    local
rpool/fs1  canmount    noauto      local
rpool/fs1  mounted     no
```

Comment (don-brady) [reply cmd/zfs/zfs_main.c]: Added a check for `zfs_is_mounted()` so that we don't export an unmounted dataset.

File: cmd/zfs/zfs_main.c [hunk 2]
```diff
@@ -6549,8 +6549,13 @@ share_mount_one(zfs_handle_t *zhp, int op, int flags, char *protocol,
 		    "'canmount' property is set to 'off'\n"), cmdname,
 		    zfs_get_name(zhp));
 		return (1);
-	} else if (canmount == ZFS_CANMOUNT_NOAUTO && !explicit) {
-		return (0);
+	} else if (canmount == ZFS_CANMOUNT_NOAUTO) {
+		if (generate) {
```
Comment (sebroy) [inline cmd/zfs/zfs_main.c]: I'd love to see some comments here explaining what the logic represents. It's not helping me that there is no comment explaining what `generate` and `explicit` mean.

Comment (don-brady) [reply cmd/zfs/zfs_main.c]: added some comments...

**Pull request merged**